src/dune_pkg/fetch.ml

```diff
@@ -260,6 +261,149 @@ let fetch ~unpack ~checksum ~target (url : OpamUrl.t) =
+  let make_stdout () = Process.Io.make_stdout ~output_on_success:Swallow ~output_limit
+  let make_stderr () = Process.Io.make_stderr ~output_on_success:Swallow ~output_limit
+
+  let run { git; dir } =
+    let stdout_to = make_stdout () in
+    let stderr_to = make_stderr () in
+    Process.run ~dir ~display ~stdout_to ~stderr_to failure_mode git
+  ;;
+
+  let run_capture_line { git; dir } =
+    Process.run_capture_line ~dir ~display failure_mode git
+  ;;
+
+  let run_capture_lines { git; dir } =
+    Process.run_capture_lines ~dir ~display failure_mode git
+  ;;
+
+  let run_capture_zero_separated_lines { git; dir } =
+    Process.run_capture_zero_separated ~dir ~display failure_mode git
+  ;;
+
+  let show { git; dir } (Rev rev) path =
+    let open Fiber.O in
+    let failure_mode = Dune_vcs.Vcs.git_accept () in
+    let command = [ "show"; sprintf "%s:%s" rev (Path.to_string path) ] in
+    let stderr_to = make_stderr () in
+    let+ res =
+      Process.run_capture_lines ~dir ~display ~stderr_to failure_mode git command
+    in
+    match res with
+    | Error _n -> None
+    | Ok v -> Some v
+  ;;
+
+  let load ~git ~dir = { git; dir }
+
+  let create ~git ~dir =
+    let open Fiber.O in
+    Path.mkdir_p dir;
+    let t = load ~git ~dir in
+    let+ () = run t [ "init"; "--bare" ] in
+    t
+  ;;
+
+  let load_or_create ~git ~dir =
+    match Path.exists dir with
+    | true -> load ~git ~dir |> Fiber.return
+    | false -> create ~git ~dir
+  ;;
+
+  module Remote = struct
+    type nonrec t =
+      { repo : t
+      ; handle : string
+      }
+
+    let head_branch = Re.(compile (seq [ str "HEAD branch: "; group (rep1 any); eol ]))
+    let update { repo; handle } = run repo [ "fetch"; handle; "--no-tags" ]
+
+    let default_branch { repo; handle } =
+      let open Fiber.O in
+      let+ contents = run_capture_lines repo [ "remote"; "show"; handle ] in
+      List.find_map contents ~f:(fun line ->
+        Re.exec_opt head_branch line
+        |> Option.map ~f:(fun groups -> Re.Group.get groups 1))
+    ;;
+
+    let equal { repo; handle } t = equal repo t.repo && String.equal handle t.handle
+
+    module At_rev = struct
+      type nonrec t =
+        { remote : t
+        ; revision : rev
+        }
+
+      let content { remote = { repo; handle = _ }; revision } path =
+        let open Fiber.O in
+        let+ lines = show repo revision path in
+        Option.map lines ~f:(String.concat ~sep:"\n")
+      ;;
+
+      let directory_entries { remote = { repo; handle = _ }; revision = Rev rev } path =
+        let open Fiber.O in
+        (* TODO future: cache ls-tree *)
+        let+ all_files =
+          run_capture_zero_separated_lines
+            repo
+            [ "ls-tree"; "-z"; "--name-only"; "-r"; rev ]
+        in
+        List.filter_map all_files ~f:(fun entry ->
+          let path_entry = Path.of_string entry in
+          Path.descendant path_entry ~of_:path)
```

Comment:
is it possible for this function to return `None`?

Reply:
Yes, this iterates over all the files in the repo, so also files like `repo`, for which this would return `None`.